### PR TITLE
[Ready for Review] GUI - Login with GitHub

### PR DESCRIPTION
**Pull request by arnohovhannisyan**
#### Short description of what this resolves:
This PR allows the user to log in to GitHub using a button, and the extension will automatically generate a token with the correct scopes and find the gist id.

#### Changes proposed in this pull request:

- Login to GitHub using OAuth
- Create a GUI for Settings and Welcome page
- Allow user to choose from a list of available gists

**Fixes**: #506

#### How Has This Been Tested?

I have tested this by logging in to GitHub and checking if the token and gist id has been saved properly.

#### Screenshots (if appropriate):
Landing Page
Settings Page
Gist Selection Page

#### Checklist:

- [x] I have read the [contribution](https://github.com/shanalikhan/code-settings-sync/blob/master/CONTRIBUTING.md#setup-extension-locally) guidelines.
- [x] My change requires a change to the documentation and GitHub Wiki.
- [x] I have updated the documentation and Wiki accordingly.

#### Status:
- [x] Authenticate with OAuth
- [x] Save Token
- [x] GUI for Gist Selection page
- [x] GUI for Settings page
- [x] GUI for Welcome page

### CODE REVIEW

**Comment by arnohovhannisyan** [comment]
@shanalikhan As you can see by the title, this PR is ready for extensive review. I just wanted to bring this to your attention because it has been a while since the release of `v3.2.9`.

**Comment by shanalikhan** [comment]
Yes i know, give me some time. I will provide a review in detail.

**Comment by shanalikhan** [comment]
Here is the original UI layout:
https://github.com/shanalikhan/code-settings-sync/pull/825#issuecomment-480175040
I am using the check box to make sure the the status.
You can copy this checklist and post in later posts to show the status.

**Overall:**
- [ ] It should be the same font / style the image i posted. Reference : [Image](https://user-images.githubusercontent.com/8774556/55582956-2612d980-573a-11e9-8487-f322b68d0b48.png)

**Title**

- [ ] Title as same color, size and font with the logo. 
- [ ] The Image in Title should be large like the Bookmark's ribbon logo and should be on middle.
- [ ] Settings Sync text size should be small compared to cloud image but should be placed in the middle of the cloud - See the image ribbon and Bookmarks place over there.
- [ ] Transparent logo should be placed on repo as another file ( we dont want to use the external link )
- [ ] The Text contrast should be aligned with the background e.g Black or Grey Text on White Background , White Text on black background or any other color to achieve the contrast ratio.


**Description:**
- [x] Description should be "Synchronize Settings, Snippets, Themes, File Icons, Launch, Keybindings, Workspaces and Extensions Across Multiple Machines Using GitHub Gist"
- [x] Below to description , please add "By @shanalikhan"
- [ ] Description should be same color, size and font as above image.

**Two Columns**
- [x] There should be two columns.
- [ ] Both columns should be scroll able.
- [ ] Height of those columns should be dynamic and then apply scroll bar when more content is available. Height should be till bottom while keeping the space for 'Help` and `Sponsor` section.
- [ ] Both columns should have white text or any text while keeping the contrast with the background.
- [ ] Links can be blue , prominent

**Left Column**

- [ ] "Whats new in Settings Sync" should be on center on left column. It should have same font size, color as in the reference image for bookmarks.
- [x] Use same green button for "New", light red for "Fix"

**Right Column**
- [ ] "Configurations" needs to have same size and style as reference image.
- [x] Some description like "Login via github to setup the Settings Sync or change configurations manually"
- [x] There should be two buttons after description
- [ ] Buttons should be in one row , left and right
- [x] Title "Show Your Support" 
- [x] Description
- [x] Two buttons Payment and sponsor
- [ ] Two buttons should be one row, left and right.
- [ ] Add Link
    - Write a Review
    - Star on Github
    - Follow me on Twitter



**Below to columns**
**Left**
- [x] "Need Help ?"
- [ ] It should have same color and style

**Right**
- [x] Add Sponsors section
- [x] Description for Sponsors "Settings Sync Looking for sponsors to display here. Contact me on twitter or github."

**Edit Configuration Page**

- [x] It will a two columns, the title, description and all things will be the stay on same location.
- [x] Only welcome page 2 columns and below it should be replaced with configuration page.
- [ ] Please fix the description section to show that same title and description 

- [x] Two Scroll able Columns


**Left Column**
- [x] Title "Environment Settings"
- [ ] Center Aligned and same font name and color, size as above reference.

Add text boxes and checkboxes for each of the Settings Sync config present in settings.json.
Below to each of the configuration textbox, put settings description present in localization files.

Right Column - Scrollable
"Title - Global Settings"
All the custom json settings should be available like in above picture.

After these two columns.
There will be "Save" button in right side- When user clicks on save. Text should appear on left to the button that settings has been saved.

Goal: It should use same font and styling like the image i posted.
Please make sure, the font size and page layout should be responsive and must align with the screen size keep font ratio with screen size intact.

Once the welcome and configuration page layout is completed, i will move to code review and github sync process.

Thanks

**Comment by arnohovhannisyan** [comment]
**Overall:**

* [x]  It should be the same font / style the image i posted. Reference : [Image](https://user-images.githubusercontent.com/8774556/55582956-2612d980-573a-11e9-8487-f322b68d0b48.png)

**Title**

* [ ]  Title as same color, size and font with the logo.
* [ ]  The Image in Title should be large like the Bookmark's ribbon logo and should be on middle.
* [ ]  Settings Sync text size should be small compared to cloud image but should be placed in the middle of the cloud - See the image ribbon and Bookmarks place over there.
* [ ]  Transparent logo should be placed on repo as another file ( we dont want to use the external link )
* [ ]  The Text contrast should be aligned with the background e.g Black or Grey Text on White Background , White Text on black background or any other color to achieve the contrast ratio.

**Description:**

* [x]  Description should be "Synchronize Settings, Snippets, Themes, File Icons, Launch, Keybindings, Workspaces and Extensions Across Multiple Machines Using GitHub Gist"
* [x]  Below to description , please add "By @shanalikhan"
* [x]  Description should be same color, size and font as above image.

**Two Columns**

* [x]  There should be two columns.
* [x]  Both columns should be scroll able.
* [ ]  Height of those columns should be dynamic and then apply scroll bar when more content is available. Height should be till bottom while keeping the space for 'Help`and`Sponsor` section.
* [x]  Both columns should have white text or any text while keeping the contrast with the background.
* [x]  Links can be blue , prominent

**Left Column**

* [x]  "Whats new in Settings Sync" should be on center on left column. It should have same font size, color as in the reference image for bookmarks.
* [x]  Use same green button for "New", light red for "Fix"

**Right Column**

* [x]  "Configurations" needs to have same size and style as reference image.
* [x]  Some description like "Login via github to setup the Settings Sync or change configurations manually"
* [x]  There should be two buttons after description
* [ ]  Buttons should be in one row , left and right
* [x]  Title "Show Your Support"
* [x]  Description
* [x]  Two buttons Payment and sponsor
* [x]  Two buttons should be one row, left and right.
* [x]  Add Link
  
  * Write a Review
  * Star on Github
  * Follow me on Twitter

**Below to columns**
**Left**

* [x]  "Need Help ?"
* [x]  It should have same color and style

**Right**

* [x]  Add Sponsors section
* [x]  Description for Sponsors "Settings Sync Looking for sponsors to display here. Contact me on twitter or github."

**Edit Configuration Page**

* [x]  It will a two columns, the title, description and all things will be the stay on same location.
* [x]  Only welcome page 2 columns and below it should be replaced with configuration page.
* [x]  Please fix the description section to show that same title and description
* [x]  Are these two columns are scroll able.

**Left Column**

* [x]  Title "Environment Settings"
* [x]  Center Aligned and same font name and color, size as above reference.
* [x]  Add textboxes and checkbox for each configuration present in `settings.json`
* [x]  Put some padding in the checkboxes as currently they are all merged.

**Right Column**

* [x]  Title - Global Settings
* [x]  All the custom json settings should be available like in above picture.
* [x]  Saving automatically.

Conclusion:

* [x]  We need to make sure the font color, style should be the same as reference image.
* [ ]  Page background is transparent, can be changed by theme settings.
* [ ]  Once the welcome and configuration page layout is completed, i will move to code review and github sync process.

Preview:
Landing
Settings

**Comment by shanalikhan** [comment]
Let me know once you complete all the checklist so i will start the UI review.

**Comment by arnohovhannisyan** [comment]
@shanalikhan I have completed your checklist.

**Comment by shanalikhan** [comment]
**Review** on the pending tasks from https://github.com/shanalikhan/code-settings-sync/pull/876#issuecomment-490395005

**Overall:**
- [ ] It should be the same font / style the image i posted. Reference : [Image](https://user-images.githubusercontent.com/8774556/55582956-2612d980-573a-11e9-8487-f322b68d0b48.png)

**Title**

- [x] Title as same color, size and font with the logo. 
- [ ] Settings Sync text size should be small compared to cloud image but should be placed in the middle of the cloud - See the image ribbon and Bookmarks place over there.
    - [ ] Can you make something like in shown image below as left to right, Text should be middle to image - I have made a similar image you can see in bottom.
- [x] Transparent logo should be placed on repo as another file ( we dont want to use the external link )
- [ ] The Text contrast should be aligned with the background e.g Black or Grey Text on White Background , White Text on black background or any other color to achieve the contrast ratio.
    - [ ] See the reference image

**Description:**
- [ ] Description should be same color, size and font as above image.
    - [ ] Text should be center aligned, currently its left aligned. Same for the "By shanalikhan" in next line.

**Two Columns**
- [x] Both columns should be scroll able.
- [x] Height of those columns should be dynamic and then apply scroll bar when more content is available. Height should be till bottom while keeping the space for 'Help` and `Sponsor` section.
- [x] Both columns should have white text or any text while keeping the contrast with the background.
- [x] Links can be blue , prominent

**Left Column**

- [ ] "Whats new in Settings Sync" should be on center on left column. It should have same font size, color as in the reference image for bookmarks.
    - [ ] Add some margin between the "Whats New" and release notes.

**Right Column**
- [x] "Configurations" needs to have same size and style as reference image.
- [x] Buttons should be in one row , left and right
- [x] Two buttons should be one row, left and right.
    - [ ] Button when clicked keep their color unchanged, we need to use hover and simple hyperlink only.
- [x] Add Link
    - Write a Review
    - Star on Github
    - Follow me on Twitter



**Below to columns**
**Left**
- [x] It should have same color and style

**Edit Configuration Page**
- [x] Please fix the description section to show that same title and description 

- [x] Are these two columns are scroll able.


**Left Column**
- [x] Center Aligned and same font name and color, size as above reference.
- [x] Put some padding in the checkboxes as currently they are all merged.



**Right Column** 
- [ ] Title - Global Settings
- [ ] Order the text boxes by 
    - Token
    - Gist name
    - Enterprise URL
    - Ignored Folders
    - Ignored Extensions
    - Host name
    - Ignored Files
    - etc

Conclusion:
- [ ] We need to make sure the font color, style should be the same as reference image.
- [x] Page background is transparent, can be changed by theme settings.
- [ ] Once the welcome and configuration page layout is completed, i will move to code review and github sync process.


image


image


**Additional Discussion:**

When the settings are opened, and use selects "Sync : Reset Extension Settings" the gist and token are although cleared but it stays in the page. what do you suggest to handle this ?

**Comment by arnohovhannisyan** [comment]
@shanalikhan
> When the settings are opened, and use selects "Sync : Reset Extension Settings" the gist and token are although cleared but it stays in the page. what do you suggest to handle this ?


We can probably close the settings page and open the landing page for the user to log in.

> When selected gist is selected, can we make better UI to let user know that it has been selected and need to close the tab.

Sure, I can make some sort of modal that informs the user that the gist has been saved. It can also have a button to close the page.

**Comment by arnohovhannisyan** [comment]
@shanalikhan Once I find a way to handle resets, everything should be ready.

Landing
Settings
Gist

**Comment by arnohovhannisyan** [comment]
@shanalikhan Ok, I have finished everything on the checklist.

> When the settings are opened, and use selects "Sync: Reset Extension Settings" the gist and token are although cleared but it stays in the page. what do you suggest to handle this?

The settings page will automatically refresh the values when a reset is triggered.

**Comment by shanalikhan** [comment]
sure, i will start a code review.
I am thinking to release this as a part of `v3.4.0`
Because [v3.3.0](https://github.com/shanalikhan/code-settings-sync/milestone/30) already has alot of breaking changes like OSS fix. I believe if some case is left behind on that PR, alot of issues will open and i dont want this PR to add more issues list ( if it arises any )

What do you suggest ?

**Comment by arnohovhannisyan** [comment]
@shanalikhan Sure, we can wait to release it in v3.4.0.

**Comment by shanalikhan** [comment]
Great.
if v3.3.0 run smoothly without any release of v3.3.1. We will release `v3.4.0` two weeks after.

**Comment by shanalikhan** [comment]
Can you fix the conflicts so i can start code review and merge for upcoming release

**Comment by arnohovhannisyan** [comment]
@shanalikhan Sure.

**Comment by arnohovhannisyan** [comment]
@shanalikhan The merge conflicts have been resolved.

**Comment by shanalikhan** [comment]
Can you add the fields for the recently added configurations in `v3.3.0` and `v3.4.0`

**Comment by arnohovhannisyan** [comment]
@shanalikhan Ok, I have added the new configuration items to the settings page.

**Comment by shanalikhan** [comment]
Some contrast config are changed.

The page doesnt match with the theme background now.

image

Comparing with image
https://github.com/shanalikhan/code-settings-sync/pull/876#issuecomment-496414995


Also check the contrast ratio of font

image

**Comment by arnohovhannisyan** [comment]
@shanalikhan 

> The page doesn't match with the theme background now.

Are you sure we should match the theme background? It would be much harder to guarantee good contrast. A great example of this is the screenshot of the Bookmarks extension that you attached. The links, buttons, and images all have very low contrast and are hard to see.

Therefore I suggest we leave it like it currently is (dark on dark themes, light on light themes). However, if it's absolutely necessary, I can update it to match the background (with the potential side effect of bad contrast). What are your thoughts?

File: src/service/github.oauth.service.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  public async StartProcess() {
+    this.server = this.app.listen(this.port);
+    this.app.get("/callback", async (req, res) => {
+      const params = new URLSearchParams(
+        await (await this.getToken(req.param("code"))).text()
+      );
+
+      res.send("<script>window.close();</script>");
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: in Firefox new tab didnt close ? Can you look into it.
As we need to close the tabs in all browsers.

File: src/service/github.oauth.service.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      const user = await this.getUser(token);
+
+      const gists = await this.getGists(token, user);
+      state.commons.webviewService.OpenGistSelectionpage(gists);
+    });
+  }
+
+  public getToken(code: string) {
+    const params = new URLSearchParams();
+    params.append("client_id", "cfd96460d8b110e2351b");
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: is there any way we can hide this? What are your thoughts

File: src/service/github.oauth.service.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    params.append("client_secret", "ed46bd3a0f736e0da57308e86ca5fa3cf8688582");
+    params.append("code", code);
+
+    return fetch("https://github.com/login/oauth/access_token", {
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: `https://github.com/login/oauth/access_token`

What about github enterprise login defined in Global Settings.
We need to create url depending on default github or user github enterprise login and make api calls over there instead of hard coding here.

File: src/service/github.oauth.service.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      method: "POST",
+      body: params
+    });
+  }
+
+  public async getGists(token: string, user: string) {
+    const res = await fetch(`https://api.github.com/users/${user}/gists`, {
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: Github enterprise support

File: src/service/github.oauth.service.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  public async saveToken(token: string) {
+    const currentSettings = await state.commons.GetCustomSettings();
+    currentSettings.token = token;
+    state.commons.SetCustomSettings(currentSettings);
+  }
+
+  public async getUser(token: string) {
+    const res = await fetch("https://api.github.com/user", {
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: Github enterprise support

File: src/service/github.oauth.service.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import * as express from "express";
+import { Server } from "http";
+import fetch from "node-fetch";
+import { URLSearchParams } from "url";
+import { state } from "../state";
+
+export class GitHubOAuthService {
```
Comment (shanalikhan) [inline src/service/github.oauth.service.ts]: how are we handling exceptions in this entire class ?
We need to throw manual exceptions like other chunks in repo.

File: src/service/webview.service.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+import { ExtensionConfig } from "../models/extensionConfig.model";
+import { UISettingType } from "../models/settingType.model";
+import { IWebview } from "../models/webview.model";
+import { state } from "../state";
+import { GitHubOAuthService } from "./github.oauth.service";
+
+export class WebviewService {
+  private globalSettings = [
+    {
+      name: "Access Token",
```
Comment (shanalikhan) [inline src/service/webview.service.ts]: We need to localize these names and placeholders.

File: src/service/webview.service.ts [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+      name: "Open Token Link",
+      placeholder: "",
+      type: UISettingType.Checkbox,
+      correspondingSetting: "openTokenLink"
+    }
+  ];
+
+  private environmentSettings = [
+    {
+      name: "Gist ID",
```
Comment (shanalikhan) [inline src/service/webview.service.ts]: localization...

File: src/service/webview.service.ts [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+      }
+    );
+    landingPanel.webview.onDidReceiveMessage(async message => {
+      switch (message.command) {
+        case "loginWithGitHub":
+          new GitHubOAuthService(54321).StartProcess();
+          vscode.commands.executeCommand(
+            "vscode.open",
+            vscode.Uri.parse(
+              "https://github.com/login/oauth/authorize?scope=gist%20read:user&client_id=cfd96460d8b110e2351b&redirect_uri=http://localhost:54321/callback"
```
Comment (shanalikhan) [inline src/service/webview.service.ts]: github enterprise support.

File: README.md [hunk 10]
```diff
@@ -69,75 +70,35 @@ All extensions and complete User Folder that Contains
...
-![github account access token](https://shanalikhan.github.io/img/upload1.png)
-
-**Upload your settings automatically and the extension gives you Gist ID in the system message.**
-Gist ID is needed to access the data you have uploaded with your token. Copy this Gist ID in order to download the settings to other machines.
-
-![uploaded automatically](https://shanalikhan.github.io/img/upload2.png)
+When downloading or uploading for the first time, the welcome page will automatically open, where you can log in with GitHub.
 
-You can always **verify created gist** on the following url:
+![Login with GitHub](https://user-images.githubusercontent.com/25834068/55663873-93c21280-57d9-11e9-8c24-7088a81ed47c.png)
```
[the remark this answers is not in the document]
Comment (arnohovhannisyan) [reply README.md]: Sure, I'll do that.

File: src/service/github.oauth.service.ts (hunk 2 above)
Comment (arnohovhannisyan) [reply src/service/github.oauth.service.ts]: As far as I know, there is no way to hide the client id/secret in an extension like this.

File: src/service/github.oauth.service.ts (hunk 1 above)
Comment (arnohovhannisyan) [reply src/service/github.oauth.service.ts]: I wasn't able to get this to work because of limitations from modern browsers, but I have made a simple screen informing the user that they may close the tab.

image

File: src/service/github.oauth.service.ts (hunk 3 above)
Comment (arnohovhannisyan) [reply src/service/github.oauth.service.ts]: Done

File: src/service/github.oauth.service.ts (hunk 6 above)
Comment (arnohovhannisyan) [reply src/service/github.oauth.service.ts]: If an error occurs during the `fetch` to GitHub (most likely because of invalid URL), `Commons.LogException` is called with the error and a possible cause ("Sync: Invalid GitHub Enterprise URL."). Otherwise, the error is logged with `Commons.LogException` using the generic error message, informing the user to check the console.

File: src/service/github.oauth.service.ts (hunk 1 above)
Comment (shanalikhan) [reply src/service/github.oauth.service.ts]: sounds good!

File: src/service/github.oauth.service.ts (hunk 2 above)
Comment (shanalikhan) [reply src/service/github.oauth.service.ts]: Alright, then I think we can keep it.

File: release-notes.json [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+      "type": "NEW",
+      "color": "success",
+      "author": "knyhle",
+      "pullRequest": "828"
+    },
+    {
+      "details": "Update <b>Documentation</b>",
+      "type": "NEW",
+      "color": "success",
+      "author": "colinaaa",
```
Comment (shanalikhan) [inline release-notes.json]: `author` and `pullRequest` supports empty string, right ?
Because for some issues I often do commit rather PR.

Comment (arnohovhannisyan) [reply release-notes.json]: I have just updated it to do so. You can either leave out those options or set them to empty strings to not show them.

File: src/commons.ts [hunk 12]
```diff
@@ -106,47 +108,15 @@ export default class Commons {
...
-        }
-        const tokTemp: string = await this.GetTokenAndSave(cusSettings);
-        if (!tokTemp) {
-          const msg = localize("common.error.tokenNotSave");
-          vscode.window.showErrorMessage(msg);
-          throw new Error(msg);
-        }
-        cusSettings.token = tokTemp;
-      }
+    if (cusSettings.token === "" || extSettings.gist === "") {
```
Comment (shanalikhan) [inline src/commons.ts]: If you explore the process of downloading public gist of another user. 
It doesnt ask the token for downloading.
So need to check if the public gist is enabled it should not check for token.

We need to make it supported

File: src/commons.ts [hunk 13]
```diff
@@ -106,47 +108,15 @@ export default class Commons {
...
-    askGist: boolean
-  ): Promise<LocalConfig> {
+  public async InitalizeSettings(): Promise<LocalConfig> {
     const settings = new LocalConfig();
     const extSettings = this.GetSettings();
     const cusSettings = await this.GetCustomSettings();
 
-    if (cusSettings.token === "") {
-      if (askToken === true) {
-        askToken = !cusSettings.downloadPublicGist;
```
Comment (shanalikhan) [inline src/commons.ts]: here is the check for `downloadPublicGist`

File: README.md [hunk 14]
```diff
@@ -212,7 +175,6 @@ You can customize the settings in gist settings like:
     "sync.forceDownload": false,
     "sync.forceUpload": false,
     "sync.quietSync": false,
-    "sync.askGistName": false,
```
Comment (shanalikhan) [inline README.md]: Why is this removed?
If its `enabled` it asks for GIST Name.

File: ui/landing-page/landing-page.html [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta
+      name="viewport"
+      content="width=device-width, initial-scale=1, shrink-to-fit=no"
+    />
+    <link
+      rel="stylesheet"
+      href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
```
Comment (shanalikhan) [inline ui/landing-page/landing-page.html]: What do you suggest, i think we should use the `js` and `css` locally instead of link.
If internet is not working the page UI wont show up properly.

File: ui/settings/settings.html [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <meta
+      name="viewport"
+      content="width=device-width, initial-scale=1, shrink-to-fit=no"
+    />
+    <link
+      rel="stylesheet"
+      href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
```
Comment (shanalikhan) [inline ui/settings/settings.html]: should be all local `js` and `css`

File: README.md (hunk 14 above)
Comment (arnohovhannisyan) [reply README.md]: This setting exists in `CustomConfig`, not `ExtensionConfig`. Therefore having this setting mentioned in the `README.md` could mislead people.

Comment (shanalikhan) [reply README.md]: Nice catch, thanks

File: ui/landing-page/landing-page.html (hunk 15 above)
Comment (arnohovhannisyan) [reply ui/landing-page/landing-page.html]: I have added all the required files to the `ui/shared/vendor` folder.

File: src/commons.ts (hunk 12 above)
Comment (arnohovhannisyan) [reply src/commons.ts]: Ok, my new commit supports public gists

**Comment by shanalikhan** [comment]
Does `webpack` minify the `html`, `js` and `css` and all other new files UI?

**Comment by arnohovhannisyan** [comment]
@shanalikhan The css and js files are already minified, while the fonts can't be because they're binaries. Only the HTML files are left without minification, but it's probably fine because they are pretty small and minification wouldn't have a big difference on their file size.

**Comment by shanalikhan** [comment]
I have merged the changes.
It works on debug mode
When i tried creating package using `vsce package` and install the file by command line and run the code.

Following error occurred when clicked on Advance Settings

image


Can you test your scenarios by creating the package file also and send me PR ?

**Comment by arnohovhannisyan** [comment]
@shanalikhan I have figured out the problem, but only a temporary solution. This is caused by `webpack-node-externals`, which I added to remove the warning in webpack. However, this is preventing a lot of dependencies from being bundled. After an hour of searching, the only way I could come up with was to stop using `webpack-node-externals` and just disable warnings. What do you think?

**Comment by shanalikhan** [comment]
image

Currently the process to download the Public Gist has become very hard.

Can you add a button below to Login with Github as "Download Public Gist" - Style should be like hyperlinks (not buttons). When user click that button. Old Input box should be provided to write the Gist ID to configure as public gist.
